File: Tigram/Controllers/ThemesViewController.m
//
//  ThemesViewController.m
//  Tigram
//
//  Created by Маргарита Коннова on 03/03/2019.
//  Copyright © 2019 Margarita Konnova. All rights reserved.
//

#import "ThemesViewController.h"
#import "ThemesViewControllerDelegate.h"

@implementation ThemesViewController

// MARK: LifeCycle of ThemesViewController
- (void) viewDidLoad {
    [super viewDidLoad];
    
    NSLog(@"ThemesViewController: Objective-C");
    // Default - white color
    // Don't need to call release fot it
    UIColor * lightColor = [UIColor whiteColor];
    UIColor * darkColor = [[UIColor alloc] initWithRed:83 / 256.0 green:103 / 256.0 blue:120 / 256.0 alpha:1.0];
    UIColor * champagneColor = [[UIColor alloc] initWithRed:244 / 256.0 green:217 / 256.0 blue:73 / 256.0 alpha:1.0];
    
    _model = [[Themes alloc] initThemesWithFirstColor:lightColor secondColor:darkColor thirdColor:champagneColor];
    
    [darkColor release];
    [champagneColor release];
}

- (void) viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    
    NSUserDefaults * data = [NSUserDefaults standardUserDefaults];
    NSString * themeNameForBackground = [data objectForKey:@"Theme"];
    if ([themeNameForBackground isEqual: @"light"]) {
        self.view.backgroundColor = [self.model light];
    }
    else if ([themeNameForBackground isEqual: @"dark"]) {
        self.view.backgroundColor = [self.model dark];
    }
    else if ([themeNameForBackground isEqual: @"champagne"]) {
        self.view.backgroundColor = [self.model champagne];
    }
}

- (void) dealloc {
    // Reduces the number of references for _model by one to avoid memory leak
    [_model release];
    // Sets _model pointer to nil to avoid dangling pointers
    _model = nil;
    // Sets _delegate pointer to nil to avoid dangling pointers
    _delegate = nil;
    
    [super dealloc];
}

// MARK: Getters for properties
- (Themes *) model {
    return _model;
}

- (id <ThemesViewControllerDelegate>) delegate {
    return _delegate;
}

// MARK: Setters for properties
- (void) setModel:(Themes *)model {
    if (_model != model) {
        // Reduces the number of references for previous model by one to avoid memory leak
        [_model release];
        // Increase the number of references for new model by one
        _model = [model retain];
    }
}

- (void) setDelegate:(id <ThemesViewControllerDelegate>)delegate {
    // Don't need to call retain because of using assign attribute
    if (_delegate != delegate)
        _delegate = delegate;
}

// MARK: User interaction
- (IBAction) closeButtonClicked:(UIButton *)sender {
    // Returns to ConversationsListViewController
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (IBAction) lightThemeButtonClicked:(UIButton *)sender {
    self.view.backgroundColor = [self.model light];
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    [userDefaults setObject:@"light" forKey:@"Theme"];
    [_delegate themesViewController:self didSelectTheme:[self.model light]];
}

- (IBAction) darkThemeButtonClicked:(UIButton *)sender {
    self.view.backgroundColor = [self.model dark];
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    [userDefaults setObject:@"dark" forKey:@"Theme"];
    [_delegate themesViewController:self didSelectTheme:[self.model dark]];
}

- (IBAction) champagneThemeButtonClicked:(UIButton *)sender {
    self.view.backgroundColor = [self.model champagne];
    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
    [userDefaults setObject:@"champagne" forKey:@"Theme"];
    [_delegate themesViewController:self didSelectTheme:[self.model champagne]];
}

@end

File: Tigram/Controllers/ThemesViewController.swift
//
//  ThemesViewController.swift
//  Tigram
//
//  Created by Маргарита Коннова on 05/03/2019.
//  Copyright © 2019 Margarita Konnova. All rights reserved.
//

import UIKit

// For compatibility with Objective-C class
protocol ThemesViewControllerDelegate {
     func themesViewController(_ controller: ThemesViewController!, didSelectTheme selectedTheme: UIColor!)
}


class ThemesViewController: UIViewController {

    // For compatibility with Objective-C class
    var delegate: UIViewController?
    var closureForSettingNewTheme: ((_ color: UIColor?, _ viewController: UIViewController?) -> Void)?
    var model: Themes?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        NSLog("ThemesViewController: Swift")

        let lightColor: UIColor = .white
        let darkColor: UIColor = UIColor.init(red: 83 / 256.0, green: 103 / 256.0, blue: 120 / 256.0, alpha: 1.0)
        let champagneColor: UIColor = UIColor.init(red: 244 / 256.0, green: 217 / 256.0, blue: 73 / 256.0, alpha: 1.0)
        model = Themes(themesWithFirstColor: lightColor, secondColor: darkColor, thirdColor: champagneColor)
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        let themeName = UserDefaults.standard.string(forKey: "Theme")
        switch themeName {
        case "light":
            self.view.backgroundColor = UIColor.white
        case "dark":
            // let darkColorForTheme = UIColor.init(red: 83 / 256.0, green: 103 / 256.0, blue: 120 / 256.0, alpha: 1.0)
            self.view.backgroundColor = model?.dark
        case "champagne":
            // let champagneColorForTheme = UIColor.init(red: 244 / 256.0, green: 217 / 256.0, blue: 73 / 256.0, alpha: 1.0)
            self.view.backgroundColor = model?.champagne
        default:
            NSLog("No valid name for theme")
        }
    }
    
    // Returns user to conversations list screen
    @IBAction func closeButtonClicked(_ sender: Any) {
        dismiss(animated: true, completion: nil)
    }
    
    @IBAction func lightThemeButtonClicked(_ sender: Any) {
        closureForSettingNewTheme?(model?.light, self)
        UserDefaults.standard.set("light", forKey: "Theme")
    }
    
    @IBAction func darkThemeButtonClicked(_ sender: Any) {
        closureForSettingNewTheme?(model?.dark, self)
        UserDefaults.standard.set("dark", forKey: "Theme")
    }
    
    @IBAction func champagneThemeButtonClicked(_ sender: Any) {
        closureForSettingNewTheme?(model?.champagne, self)
        UserDefaults.standard.set("champagne", forKey: "Theme")
    }
}
